### PR TITLE
Frames as Video

### PR DESCRIPTION
**Pull request by sashankaryal**
This PR adds support for rendering ordered dynamic groups of images as a video.
videoframes-small

### Testing the feature
1. You may get started with any image dataset that you can create an ordered dynamic group out of. It might also be easier to take a video dataset, run `to_frames()` view stage on it, and create an ordered dynamic group out of that.
2. After creating an ordered dynamic group, open the modal and select "video" from display options for dynamic groups navigation.

### Files to pay attention to
This is a list of files that have the most significance.
1. `packages/looker/src/elements/imavid/index.ts`
2. `packages/looker/src/lookers/imavid/controller.ts`
3. `packages/looker/src/lookers/imavid/buffer-manager/index.ts`
4. `packages/looker/src/lookers/imavid/store.ts`
5. `packages/looker/src/lookers/imavid/ima-vid-frame-samples.ts`

### Todos
- [ ] E2E for golden path
- Known issues and bugs
  - [ ] Tagging
  - [ ] Zoom and crop are broken

### Known areas of improvement:
1. Add an option to switch to "imavid" looker from the grid. Currently, it's only available from the modal. This should be an easy add.
2. Use a better mechanism than `setTimeout()` for throttling `requestAnimationFrame()`.
3. Reconcile common elements of video looker and ima-vid looker.
4. Improve type safety of looker code.

### Testing
1. Unit tests for buffer manager `packages/looker/src/lookers/imavid/buffer-manager/buffer-manager.test.ts`
2. E2E tests for golden path forthcoming

### CODE REVIEW

File: app/packages/looker/src/lookers/imavid/buffer-manager/buffer-manager.test.ts [hunk 1]
```diff
@@ -0,0 +1,242 @@
+import { beforeEach, describe, expect, test } from "vitest";
```
Comment (benjaminpkane) [inline app/packages/looker/src/lookers/imavid/buffer-manager/buffer-manager.test.ts:1]: Awesome! These tests are super valuable. It'd be cool to apply the buffer manager to `VideoLooker` in the future 🚀

File: app/packages/looker/src/lookers/imavid/buffer-manager/index.ts [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+   * Returns the metadata for a buffer range at the specified index.
+   */
+  public getMetadataForBufferRange(rangeIndex: number) {
+    return this.bufferMetadata[rangeIndex];
+  }
+
+  /**
+   * Adds a buffer range to the buffer.
+   * If the new range overlaps with an existing range, the two ranges are merged.
+   * Time complexity: O(nlogn)
```
Comment (benjaminpkane) [inline app/packages/looker/src/lookers/imavid/buffer-manager/index.ts:52]: 🧑‍🔬

File: app/packages/looker/src/lookers/imavid/buffer-manager/index.ts [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+    // initialize a stack to store the merged ranges
+    const stack = [];
+
+    const rangesWithMetadata = [];
+    const rangesWithoutMetadata = [];
+
+    for (let i = 0; i < this.buffers.length; ++i) {
+      const range = this.buffers[i];
+
+      if (!range) {
```
Comment (benjaminpkane) [inline app/packages/looker/src/lookers/imavid/buffer-manager/index.ts:81]: Maybe we can remove the safeguards before merging? `range` is not expected to be falsey, correct?

File: app/packages/looker/src/lookers/imavid/controller.ts [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+            this.storeBufferManager.addNewRange([
+              Number(data.samples.edges[0].cursor) + 1,
+              Number(data.samples.edges[data.samples.edges.length - 1].cursor) +
+                1,
+            ]);
+          }
+          resolve();
+        },
+      });
+      // todo: see if environment.retain() is applicable here,
```
Comment (benjaminpkane) [inline app/packages/looker/src/lookers/imavid/controller.ts:234]: I think `retain()` is most valuable when subscribing to the relay data which could be done in the future, but unimportant right now.

Subscribing would mean the samples would update via their global store id if any other query/mutation/subscription updated the contents. Super valuable, but none of the other Lookers leverage it either at the moment

**Review by benjaminpkane — COMMENTED** [review]
Code looks awesome!  Can sync tomorrow about zoom/crop if you want. Can definitely help on that detail

I did not expect the sidebar to update during playback! Very cool 🥇

File: app/packages/looker/src/lookers/imavid/buffer-manager/buffer-manager.test.ts (hunk 1 above)
Comment (sashankaryal) [reply app/packages/looker/src/lookers/imavid/buffer-manager/buffer-manager.test.ts:1]: 100%

File: app/packages/looker/src/lookers/imavid/buffer-manager/index.ts (hunk 3 above)
Comment (sashankaryal) [reply app/packages/looker/src/lookers/imavid/buffer-manager/index.ts:81]: It is not... but I got at least one instance where it was falsey and I couldn't reliably reproduce it, so until I figure that out, I'll leave it there.

**Review by manivoxel51 — COMMENTED** [review]
Nice work! 🍨 

One issue running locally is JSON viewer not being rendered properly.

https://github.com/voxel51/fiftyone/assets/109545780/fe547ca5-b13b-498c-87e0-3be9abce2862

**Review by benjaminpkane — APPROVED** [review]
LGTM! 🚀

File: app/packages/looker/src/elements/common/controls.ts [hunk 5]
```diff
@@ -66,6 +66,7 @@ export class ControlsElement<
     error,
     loaded,
   }: Readonly<State>) {
+    console.log("showcontrols is ", showControls);
```
Comment (manivoxel51) [inline app/packages/looker/src/elements/common/controls.ts:69]: @sashankaryal we forgot to remove ✂️

Comment (sashankaryal) [reply app/packages/looker/src/elements/common/controls.ts:69]: Will remove it!

**Pull request merged**